From c7591deba8299bb39acd8e2efc63c28c7b23243b Mon Sep 17 00:00:00 2001
From: Juergen Gross <jgross@suse.com>
Date: Fri, 27 Sep 2019 09:00:16 +0200
Subject: xen/sched: switch struct task_slice from vcpu to sched_unit

Let the schedulers put a sched_unit pointer into struct task_slice
instead of a vcpu pointer.

Signed-off-by: Juergen Gross <jgross@suse.com>
Reviewed-by: Dario Faggioli <dfaggioli@suse.com>

diff --git a/xen/common/sched_arinc653.c b/xen/common/sched_arinc653.c
index 5cf47f5622..9ff1d7f245 100644
--- a/xen/common/sched_arinc653.c
+++ b/xen/common/sched_arinc653.c
@@ -560,9 +560,9 @@ a653sched_do_schedule(
 
     /*
      * If there are more domains to run in the current major frame, set
-     * new_task equal to the address of next domain's VCPU structure.
-     * Otherwise, set new_task equal to the address of the idle task's VCPU
-     * structure.
+     * new_task equal to the address of next domain's sched_unit structure.
+     * Otherwise, set new_task equal to the address of the idle task's
+     * sched_unit structure.
      */
     new_task = (sched_index < sched_priv->num_schedule_entries)
         ? sched_priv->schedule[sched_index].vc
@@ -598,7 +598,7 @@ a653sched_do_schedule(
      * of the selected task's VCPU structure.
      */
     ret.time = next_switch_time - now;
-    ret.task = new_task;
+    ret.task = new_task->sched_unit;
     ret.migrated = 0;
 
     BUG_ON(ret.time <= 0);
diff --git a/xen/common/sched_credit.c b/xen/common/sched_credit.c
index 350f9636fa..732fd461b5 100644
--- a/xen/common/sched_credit.c
+++ b/xen/common/sched_credit.c
@@ -1993,9 +1993,9 @@ out:
      */
     ret.time = (is_idle_vcpu(snext->vcpu) ?
                 -1 : tslice);
-    ret.task = snext->vcpu;
+    ret.task = snext->vcpu->sched_unit;
 
-    CSCHED_VCPU_CHECK(ret.task);
+    CSCHED_VCPU_CHECK(ret.task->vcpu_list);
     return ret;
 }
 
diff --git a/xen/common/sched_credit2.c b/xen/common/sched_credit2.c
index 7b0872eba5..8857c98dd8 100644
--- a/xen/common/sched_credit2.c
+++ b/xen/common/sched_credit2.c
@@ -3637,9 +3637,9 @@ csched2_schedule(
      * Return task to run next...
      */
     ret.time = csched2_runtime(ops, cpu, snext, now);
-    ret.task = snext->vcpu;
+    ret.task = snext->vcpu->sched_unit;
 
-    CSCHED2_VCPU_CHECK(ret.task);
+    CSCHED2_VCPU_CHECK(ret.task->vcpu_list);
     return ret;
 }
 
diff --git a/xen/common/sched_null.c b/xen/common/sched_null.c
index 06acaf9f90..3e6f9f72e2 100644
--- a/xen/common/sched_null.c
+++ b/xen/common/sched_null.c
@@ -819,10 +819,10 @@ static struct task_slice null_schedule(const struct scheduler *ops,
     if ( tasklet_work_scheduled )
     {
         trace_var(TRC_SNULL_TASKLET, 1, 0, NULL);
-        ret.task = idle_vcpu[cpu];
+        ret.task = idle_vcpu[cpu]->sched_unit;
     }
     else
-        ret.task = per_cpu(npc, cpu).vcpu;
+        ret.task = per_cpu(npc, cpu).vcpu->sched_unit;
     ret.migrated = 0;
     ret.time = -1;
 
@@ -857,7 +857,7 @@ static struct task_slice null_schedule(const struct scheduler *ops,
                 {
                     vcpu_assign(prv, wvc->vcpu, cpu);
                     list_del_init(&wvc->waitq_elem);
-                    ret.task = wvc->vcpu;
+                    ret.task = wvc->vcpu->sched_unit;
                     goto unlock;
                 }
             }
@@ -869,10 +869,10 @@ static struct task_slice null_schedule(const struct scheduler *ops,
             cpumask_set_cpu(cpu, &prv->cpus_free);
     }
 
-    if ( unlikely(ret.task == NULL || !vcpu_runnable(ret.task)) )
-        ret.task = idle_vcpu[cpu];
+    if ( unlikely(ret.task == NULL || !unit_runnable(ret.task)) )
+        ret.task = idle_vcpu[cpu]->sched_unit;
 
-    NULL_VCPU_CHECK(ret.task);
+    NULL_VCPU_CHECK(ret.task->vcpu_list);
     return ret;
 }
 
diff --git a/xen/common/sched_rt.c b/xen/common/sched_rt.c
index 3fbe8dad8d..856eae1faa 100644
--- a/xen/common/sched_rt.c
+++ b/xen/common/sched_rt.c
@@ -1129,7 +1129,7 @@ rt_schedule(const struct scheduler *ops, s_time_t now, bool_t tasklet_work_sched
         }
         ret.time = snext->cur_budget; /* invoke the scheduler next time */
     }
-    ret.task = snext->vcpu;
+    ret.task = snext->vcpu->sched_unit;
 
     return ret;
 }
diff --git a/xen/common/schedule.c b/xen/common/schedule.c
index 6d27c2f4a5..157b91c439 100644
--- a/xen/common/schedule.c
+++ b/xen/common/schedule.c
@@ -115,7 +115,7 @@ static struct task_slice sched_idle_schedule(
     const unsigned int cpu = smp_processor_id();
     struct task_slice ret = { .time = -1 };
 
-    ret.task = idle_vcpu[cpu];
+    ret.task = sched_idle_unit(cpu);
     return ret;
 }
 
@@ -1627,7 +1627,7 @@ static void schedule(void)
     sched = this_cpu(scheduler);
     next_slice = sched->do_schedule(sched, now, tasklet_work_scheduled);
 
-    next = next_slice.task;
+    next = next_slice.task->vcpu_list;
 
     sd->curr = next->sched_unit;
 
diff --git a/xen/include/xen/sched-if.h b/xen/include/xen/sched-if.h
index 0587a8d3d2..2b9d50b9fe 100644
--- a/xen/include/xen/sched-if.h
+++ b/xen/include/xen/sched-if.h
@@ -226,9 +226,9 @@ static inline spinlock_t *pcpu_schedule_trylock(unsigned int cpu)
 }
 
 struct task_slice {
-    struct vcpu *task;
-    s_time_t     time;
-    bool_t       migrated;
+    struct sched_unit *task;
+    s_time_t           time;
+    bool_t             migrated;
 };
 
 struct scheduler {
